File: src/main/java/baekjoon/dp/_2xN_타일링.java
package baekjoon.dp;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

/**
 * https://www.acmicpc.net/problem/11726
 *
 * **dp = 더 이상 나눌 수 없는 1단계 + 나머지 단계들**
 *
 * 할 수 있는 가장 최소 행위 : 세로로 1칸 놓기 or 가로로 2칸 놓기
 * 아래서부터 칸을 늘려가며 위의 최소행위들을 수행하게 된다
 * 즉 1칸을 늘렸을 때,
 * 세로로 1칸 추가할 때의 경우의 수(fn(n-1)) + 가로로 2칸 추가할 때의 경우의 수(fn(n-2)) 를 구하면 된다
 * 즉, fn(n) = fn(n-1) + fn(n-2)
 */

public class _2xN_타일링 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        _2xN_타일링 a = new _2xN_타일링();
        System.out.println(a.bottomUp(n));
    }

    private int mod = 10_007;

    public int bottomUp(int n) {
        if(n <= 1) {
            return 1;
        }

        int[] dp = new int[n + 1];

        dp[1] = 1;
        dp[2] = 2;

        for (int i = 3; i <= n; i++) {
            // dp[i-1], dp[i-2] 가 이미 mod 된 숫자임
            dp[i] = (dp[i - 1] + dp[i - 2]) % mod;
        }

        return dp[n];
    }

    private int[] dp;
    public int topDown(int n) {
        if(dp == null) dp = new int[n + 1];

        if(n <= 2) {
            return n;
        }

        if(dp[n] == 0) {
            dp[n] = (topDown(n - 1) + topDown(n - 2)) % mod;
        }

        return dp[n];
    }
}

File: src/main/java/baekjoon/dp/basic/2xN 타일링.kt
package baekjoon.dp.basic

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

/**
 * https://www.acmicpc.net/problem/11726
 *
 * f(n) : 2x1, 1x2로 타일을 채룰 수 있는 방법의 수
 * 끝에 놓을 수 있는 타일의 모습 : 2x1 타일 1개 놓기, 1x2 타일 2개 놓기
 * 끝에 2x1 타일을 놓았을 경우 n이 1칸 줄어들었으므로 f(n-1)
 * 끝에 1x2 타일을 놓았을 경우 n이 2칸 줄어들었으므로 f(n-2)
 * 2개를 단순히 합치면 된다(끝에 2x1타일 1개 놓았을때와 1x2 타일 2개 놓았을때의 모양은 서로 다르므로)
 *
 * 점화식 : f(n) = f(n-1) + f(n-2)
 */
fun main(args: Array<String>) {
    println(`2xN 타일링`().solution(2) == 2)
    println(`2xN 타일링`().solution(9) == 55)
    println(`2xN 타일링`().solution(1) == 1)

    try {
        val fn = `2xN 타일링`()
        val br = BufferedReader(InputStreamReader(System.`in`));
        println(fn.solution(br.readLine().toInt()))
    } catch (e: IOException) {/* */}
}

class `2xN 타일링` {
    fun solution(n: Int): Int {
        if (n <= 1) {
            return n
        }

        val dp = IntArray(n + 1)
        dp[1] = 1
        dp[2] = 2

        for (i in 3..n) {
            // 앞 숫자에 % 연산이 적용되더라도, (a + b) % c == ((a % c) + (b % c)) % c 이기 때문에 괜찮음
            dp[i] = (dp[i - 1] + dp[i - 2]) % 10_007
        }

        return dp[n]
    }
}